In [2]:
#Student ID 1002080520
import numpy as np
from numpy import loadtxt
import csv

#CODE TO TAKE THE DATA AND STORE IN THE VARIABLE

file = open('q2_cde_dataset.csv')
csvreader = csv.reader(file, delimiter = ",")
data = []
for row in csvreader:
    data.append(row)

program_data = []
for i in range(0 , len(data)):
    col= []
    for j in range(0,len(data[i])):
        if j==len(data[i])-1:
            col.append(data[i][j])
        else:
            col.append(float(data[i][j]))
    program_data.append(col)
file.close()
print("\n\nProgram Data \n\n")
print(program_data)

#We have successfully imported the data



Program Data 


[[0.088806749762916, 0.098008394324455, 0.57649058735898, 3.239527050286, ' Ceramic'], [0.078963683508211, 0.085300149997219, 0.27162441732486, 3.4442845617291, ' Metal'], [0.1317769211421, 0.070414892852672, 0.50961174676178, 3.984839303975, ' Ceramic'], [0.12238449845956, 0.081946645569603, 0.64441296678112, 3.4343857831642, ' Ceramic'], [0.11207208488382, 0.15, 0.3736066136156, 2.4684426455694, ' Plastic'], [0.11696836534332, 0.1192966406841, 0.3315496717105, 1.8135585032895, ' Plastic'], [0.12666802764538, 0.15, 0.41499442802485, 2.9562056990404, ' Plastic'], [0.10444296796273, 0.10662125348196, 0.29055615215118, 3.4594532366829, ' Plastic'], [0.13065405525064, 0.14162609670298, 0.33516279766318, 2.759442528738, ' Plastic'], [0.11931866537973, 0.10053257281629, 0.29777585695821, 4.0332186383189, ' Plastic'], [0.082429457069664, 0.03, 0.13468307216616, 1.6519180922721, ' Ceramic'], [0.091479158720172, 0.060070441860045, 0.19527332743138, 3.6252435307061, ' Metal'],

In [15]:
#Training the program data by giving data itself and predicting
Train_program_data = program_data
M_Distance =[] #Manhattan Distance
for i in range(0 , len(Train_program_data)):
    col = []
    for j in range(0,len(Train_program_data)):
        if i==j:
             j=j+1
        else:
                dist_label =[]
                dist_label.append(Train_program_data[j][4])
                X=Train_program_data[i][0]
                Y=Train_program_data[i][1]
                Z=Train_program_data[i][2]
                W=Train_program_data[i][3]
                x=Train_program_data[j][0]
                y=Train_program_data[j][1]
                z=Train_program_data[j][2]
                w=Train_program_data[j][3]
                dist = abs(X-x) + abs(Y-y) + abs(Z-z) + abs(W-w)
                dist_label.append(dist)
                col.append(dist_label)
    M_Distance.append(col)
print("\n DISTANCE MATRIX \n")
print(M_Distance)
print("\n")

In [16]:
#We have measured the distances and also labeled the points now we need to extract the distance arrange it into 
#ascending order and the again attached the labels so that we can get newly trained data set
def bubblesort(arr):   
    n = len(arr)
    # For loop to traverse through all
    # element in an array
    for i in range(n):
        for j in range(0, n - i - 1):
             
            # Range of the array is from 0 to n-i-1
            # Swap the elements if the element found
            #is greater than the adjacent element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                
only_dist =[]
for i in range(0, len(M_Distance)):
    cols=[]
    for j in range(0 , len(M_Distance[i])):
        cols.append(M_Distance[i][j][1])
    bubblesort(cols)
    only_dist.append(cols)
print(only_dist)
print(only_dist)

In [21]:
#We have arranged the distance in ascending order now we will label it
data_dist_label=[]
for i in range(0 , len(only_dist)):
    cols=[]
    for j in range(0 , len(only_dist[i])):
        for h in range(0 , len(only_dist[i])):
            if only_dist[i][j]==M_Distance[i][h][1]:
                d = M_Distance[i][h][0]
                break
        cols.append(d)
    data_dist_label.append(cols)

print(data_dist_label)

#We have successfully generated the distance label in ascending order

[[' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Cerami

In [22]:
#We will now predict by K nearest Neighbours
predict_1 = []
for i in range(0 , len(data_dist_label)):
     predict_1.append(data_dist_label[i][0])
print(predict_1)

[' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' P

In [23]:
#We will now find the accuracy for K=1
ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_1[i] == program_data[i][4]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 1 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")


Accuracy for KNN with 1 neighbour
Predicted unsuccesfully  51
Predicted Successfully  69
Accuracy 57.49999999999999%


In [24]:
a = " Metal"
b = " Plastic"
c = " Ceramic"

#We will now intialize the predictions for KNN value 3
predict_3 =[]
for i in range(0 , len(data_dist_label)):
    x=0
    y=0
    z=0
    cols=[]
    for j in range(0 , 3):
        if (str(data_dist_label[i][j]) == a):
            x=x+1
        if(str(data_dist_label[i][j])==b):
            y=y+1
        if(str(data_dist_label[i][j])==c):
            z=z+1
 
    if (x >= y) and (y >= z):
           largest = a
    elif (y >= x) and (y >= z):
        largest = b
    else:
        largest = c
    predict_3.append(largest)
print(predict_3)


[' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Plastic',

In [25]:
#Now we will find the accuracy for KNN with 3 neighbours

ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_3[i] == program_data[i][4]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 3 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")


Accuracy for KNN with 3 neighbour
Predicted unsuccesfully  58
Predicted Successfully  62
Accuracy 51.66666666666667%


In [26]:
#We will now intialize the predictions for KNN value 3
predict_5 =[]
for i in range(0 , len(data_dist_label)):
    x=0
    y=0
    z=0
    cols=[]
    for j in range(0 , 5):
        if (str(data_dist_label[i][j]) == a):
            x=x+1
        if(str(data_dist_label[i][j])==b):
            y=y+1
        if(str(data_dist_label[i][j])==c):
            z=z+1
 
    if (x >= y) and (y >= z):
           largest = a
    elif (y >= x) and (y >= z):
        largest = b
    else:
        largest = c
    predict_5.append(largest)
print(predict_5)


[' Ceramic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Metal', ' Metal', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Plastic', ' Plastic', ' Ceramic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Metal', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Ceramic', ' Ceramic', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Plastic', ' Plastic', ' Metal', ' Plastic', ' Plastic', ' Metal', ' Metal', ' Metal', ' Ceramic', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', ' Ceramic', ' Plastic', ' Metal', ' Metal', ' Plastic', ' Ceramic', ' Plastic', ' Plastic', '

In [27]:
#Now we will find the accuracy for KNN with 3 neighbours

ps=0 #predicted Successfully
pu=0 #Predicted Unsuccessfully
for i in range(0,len(program_data)):
    if predict_5[i] == program_data[i][4]:
            ps=ps+1
    else :
        pu=pu+1
accuracy_1 = ps/(ps+pu)
print("\nAccuracy for KNN with 5 neighbour")
print("Predicted unsuccesfully  "+str(pu))
print("Predicted Successfully  " + str(ps))
print("Accuracy "+str(accuracy_1*100)+"%")



Accuracy for KNN with 5 neighbour
Predicted unsuccesfully  52
Predicted Successfully  68
Accuracy 56.666666666666664%


Here we can conclude that with Manhattan Distance the accuracy is more than that of euclidian distance maximum accuracy is 56.67% in Manhattan distance where as Euclidian Distance highest accuracy was 52.5%